host: Add AVX2 support for uhd::convert

anilgurses opened the pull request:
# Pull Request Details
<!--- The title of the PR should summarize the change implemented. -->
<!--- Example commit message format: -->
<!--- `product: summary of change` -->
<!--- (leave blank) -->
<!--- `details of what/why/how an issue was addressed` -->
<!--- Keep subject lines to 50 characters, and other lines to 72 -->
<!--- characters. Refer to the "Revision Control Hygiene" section -->
<!--- CODING.md for more details. -->

## Description
<!--- Provide a general summary of your changes in the title above -->
<!--- Why is this change required? What problem does it solve? -->
AVX2 support is implemented for uhd::convert. It was previously limited with sse2. It provides performance improvements for data type conversion.

## Related Issue
<!--- Refer to any related issues here -->
<!--- If this PR fully addresses an issue, please say "Fixes #1234", -->
<!--- as this will allow Github to automatically close the related Issue -->
N/A
## Which devices/areas does this affect?
<!--- Include devices that are affected and some details on what -->
<!--- areas these changes affect, such as RF performance. -->
Affects the uhd::convert data conversion performance.

## Testing Done
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you -->
<!--- ran to see how your change affects other areas of the code, -->
<!--- etc. Then, include justifications for how your tests -->
<!--- demonstrate those affects. -->
Testing is done using the tests written previously by UHD developers. It passes all previous tests and there is no need for new tests.

## Checklist

<!--- Go over all the following points, and put an `x` in all the
<!--- boxes that apply. Note that some of these may not be valid -->
<!--- for all PRs. -->

- [x] I have read the CONTRIBUTING document.
- [x] My code follows the code style of this project. See CODING.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes, and all previous tests pass.
- [ ] I have checked all compat numbers if they need updating (FPGA compat,
      MPM compat, noc_shell, specific RFNoC block, ...)

anilgurses commented:
I have read the CLA Document and I hereby sign the CLA

anilgurses commented:
Hi! Is there anything else needed for this PR?

anilgurses commented:
If you are asking for application specific, I got fewer (relatively) overflows with srsRAN and OpenAirInterface. I don't have any quantitative results for that unfortunately. We currently use this branch and with some additional changes. If you are asking for numbers for the latest benchmark, 

```
================================================================================
                         Conversion  Best Priority      ns/sample Speedup vs Gen
--------------------------------------------------------------------------------

                  fc32 -> sc16_chdr           AVX2          0.503          2.18x
             fc32 -> sc16_item32_be           AVX2          0.505          4.05x
             fc32 -> sc16_item32_le           AVX2          0.501          2.17x
              fc32 -> sc8_item32_le           AVX2          0.431          3.08x
             sc16_item32_be -> fc32           AVX2          0.451          2.29x
             sc16_item32_be -> sc16           AVX2          0.299          2.42x
             sc16_item32_le -> fc32     SSE2/SSSE3          0.448          1.37x
             sc16_item32_le -> fc64           AVX2          0.757          1.56x
             sc16_item32_le -> sc16     SSE2/SSSE3          0.298          1.40x

```

These are the conversion types I use mostly. 

PS. I will push new changes based on your comment next week.

mbr0wn commented:
@anilgurses are you still working on this? If not, I would merge it in its current state. If you wanted to check out the alignment, then I'd wait for it.

anilgurses commented:
@mbr0wn Yes, I am. I'll push my latest changes today.

anilgurses commented:
Ok, it's done now. I also removed unaligned vs aligned handling for avx2. It seems like mm256_loadu can load aligned data almost with no penalty. Therefore, I removed it and it's much easier to understand the code. Also as a ref,
https://stackoverflow.com/questions/32612190/how-to-solve-the-32-byte-alignment-issue-for-avx-load-store-operations

anilgurses commented:
Also, I've deleted convert_benchmark until examples, which was wrong from the beginning. I've added batch functionality into converter_benchmark.cpp (let me know if it's ok structurewise). It can be executed as follows 
```
$./utils/converter_benchmark --batch --compare 
```

```
================================================================================
Summary: Fastest Converter for Each Type
================================================================================
                         Conversion  Best Priority      ns/sample Speedup vs Gen
--------------------------------------------------------------------------------
             fc32 -> sc12_item32_le     SSE2/SSSE3          0.491          2.03x
                  fc32 -> sc16_chdr           AVX2          0.239          2.91x
             fc32 -> sc16_item32_be           AVX2          0.244          6.53x
             fc32 -> sc16_item32_le           AVX2          0.252          3.01x
              fc32 -> sc8_item32_le           AVX2          0.207          4.80x
                  fc64 -> sc16_chdr           AVX2          0.342          2.93x
             fc64 -> sc16_item32_le           AVX2          0.345          3.29x
             sc12_item32_le -> fc32     SSE2/SSSE3          0.355          2.61x
             sc12_item32_le -> sc16     SSE2/SSSE3          0.176          2.76x
             sc16 -> sc12_item32_le     SSE2/SSSE3          0.253          2.15x
             sc16 -> sc16_item32_be           AVX2          0.144          2.34x
             sc16 -> sc16_item32_le           AVX2          0.141          1.31x
                  sc16_chdr -> fc32           AVX2          0.239          1.04x
                  sc16_chdr -> fc64     SSE2/SSSE3          0.462          1.33x
             sc16_item32_be -> fc32           AVX2          0.251          2.84x
             sc16_item32_be -> sc16           AVX2          0.147          2.36x
             sc16_item32_le -> fc32           AVX2          0.254          1.32x
             sc16_item32_le -> fc64           AVX2          0.447          1.33x
             sc16_item32_le -> sc16           AVX2          0.140          1.37x
              sc8_item32_le -> fc32           AVX2          0.213          2.45x
```

mbr0wn commented:
OK many thanks @anilgurses! We'll take it from here. It'll probably take a week or so until you see this PR be closed. The code on the public master could be even later than that.

anilgurses commented:
Great! Thanks for your feedback along the way.

mbr0wn commented:
> Great! Thanks for your feedback along the way.

You're very welcome. FYI, I will close this PR (even though our internal testing and review is not yet done), just so we can lock down the state of this. If we're lucky, like I said, this might make it into the next release (if not, the one after that).

Thanks again!
